docs/ldml/tr35-keyboards.md

```diff
@@ -439,8 +439,202 @@ For example, the sequence `e\m{marker}\u{300}` would be output in NFC as `è`. T
+    const noMarkers    = removeMarkers(input);
+    const markerMap    = parseMarkers(input);
+    const noMarkersNfd = normalizeNFD(noMarkers);
+    if (markerMap.empty()) {
+        // no markers in this segment.
+        return noMarkersNfd;
+    } else if(noMarkersNfd == noMarkers) {
+        // no change in this segment
+        return input;
+    } else {
+        return addBackMarkers(noMarkersNfd, markerMap);
+    }
+}
+
+// represents a removed marker, to be re-added
+interface MarkerEntry {
+    glue: string;      // 'glue' character
+    marker: Marker;  // which marker
+    processed: boolean = false; // track whether this was processed or not
+    eot: boolean = false; // track whether end of text or not
+}
+
+/** parse the string for markers */
+function parseMarkers(s: string): MarkerEntry[] {
+    // Note: parseMarkers() and removeMarkers() (not shown) could be the same function.
+    // queue of markers found
+    let lastMarkers: Marker[] = [];
+    // return value
+    let outputMap: MarkerEntry[] = [];
+
+
+    // iterate string by codepoints
+    for (i = s.begin; i< s.end;) {
+        if (markerAt(i)) {
+            // found a marker - add it to the list
+            lastMarkers.add(markerAt(i));
+            i += /* marker length */;
+        } else {
+            // From 'i' find the glue char for all markers found so far
+            // (in order)
+            // use first char of decomposed string
+            const glue = normalizeNFD(i).codePointAt(0);
+            for (m of markers) {
+                map.add({marker: m, glue: glue});
+            }
+            markers.clear();
+
+            i++; // move to next codepoint
+        }
+    }
+
+    // ran out of codepoints.
+    // any remaining markers are EOT
+    for (m of markers) {
+        map.add({marker: m, eot: true});
+    }
+}
+
+/**
+ * re-add markers to input. Mutates markerMap.
+ * @param input normalized string to re-add markers
+ * @param markerMap list of glued markers. will be mutated
+ */
+function addBackMarkers(input : string, map: MarkerEntry[]) : stringWithMarkers {
+    // quick check
+    if(map.empty()) return input;
+
+    // output string
+    let out = '';
+
+    // first, add all of the EOT entries
+    while (!map.empty() && map[map.length - 1].eot) {
+        const m = map.popEnd(); // remove the last entry
+        out = m.marker + out; // prepend the marker
+    }
+
+    // now, process input codepoints from end to beginning
+    for (i=input.end;i>=input.begin;i--) {
+        // 0. prepend this codepoint
+        out = i + out;
+
+        // 1. any markers at the end of list which match get added
+        while (!map.empty() && map[map.length - 1].glue == i) {
+            // remove last entry, so we don't have to process it any more.
+            const m = map.popEnd();
+            if (m.processed == false) {
+                out = m.marker + out; // prepend marker
+            } // else we already processed it
+        }
```

Comment:
I don't think this is entirely correct.

Also, I would like to say that use of `map` for a parameter of an array / list type has some unwanted implications that affects readability - I thought it was an actual [`Map` type](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) for a while when reviewing this section.
- Or a [Java map](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html)
- Or a [C++ map](https://cplusplus.com/reference/map/map/)
- ... you get the idea.

----

Suppose the following segment: `e` `\u{0320}` `\m{1}` `\u{0300}` `\u{0300}`
- Note:  marker in front of a non-unique member of the segment.

Based on the outer for-loop's iteration order and the inner while loop ("1."), wouldn't we get the following output?
- `e` `\u{0320}` `\u{0300}` `\m{1}` `\u{0300}`

My understanding is that this would be an incorrect result according to the spec, as the `\m{1}` should remain "attached" to the first `\u{0300}`.

Introducing something of a `.glueOrdinal` property to handle duplicated codepoints within the same segment and managing that properly would likely be sufficient... though I understand it would likely only exist to handle edge cases like this one.  This idea should viable with the current strategy:  on observing an instance of `.glue`, pre-decrement `.glueOrdinal` and then check if the resulting value matches zero.  If so, pop and insert it; if not, wait for the next `.glue` instance and break the `while` loop.

Reply:
@jahorton It's actually an array. I'll update the name.

Reply:
@jahorton 

> Suppose the following segment: `e \u{0320} \m{1} \u{0300} \u{0300}`
> .. wouldn't we get the following output?
>
> e \u{0320} \u{0300} \m{1} \u{0300}
>
> My understanding is that this would be an incorrect result according to the spec, as the \m{1} should remain "attached" to the first \u{0300}.

Good point. Thanks.